### PR TITLE
[coop] Batch of fixes for Xamarin.iOS

### PR DESCRIPTION
**Pull request by ludovic-henry**
cc @kumpera @lambdageek @rolfbjarne

### CODE REVIEW

File: mono/metadata/threads.h [hunk 1]
```diff
@@ -54,7 +54,7 @@ MONO_API void mono_threads_request_thread_dump (void);
 
 MONO_API mono_bool mono_thread_is_foreign (MonoThread *thread);
 
-extern MONO_API void mono_thread_detach_if_exiting (void);
+extern MONO_API int32_t mono_thread_detach_if_exiting (void);
```
Comment (alexrp) [inline mono/metadata/threads.h]: incorrect signature

Comment (ludovic-henry) [inline mono/metadata/threads.h]: I fixed it at https://github.com/mono/mono/pull/3074/files#diff-f1a9287559d3aa620b058619f47f00c9L1085

Comment (alexrp) [inline mono/metadata/threads.h]: Why `int32_t` when it returns a bool though?

Comment (ludovic-henry) [inline mono/metadata/threads.h]: Because we don't include `glib.h`, and the equivalent to `gboolean` is `int32_t` (https://github.com/mono/mono/blob/master/eglib/src/glib.h#L73)

Comment (alexrp) [inline mono/metadata/threads.h]: I see, it's a public header. So we should use `mono_bool` here.

File: mono/utils/mono-threads.h [hunk 2]
```diff
@@ -627,4 +627,10 @@ gboolean mono_threads_wait_pending_operations (void);
 void mono_threads_begin_global_suspend (void);
 void mono_threads_end_global_suspend (void);
 
+static inline gboolean
```
Comment (kumpera) [inline mono/utils/mono-threads.h]: Don't mark static header functions as inline. It confuses the win32 linker and any compiler under O2 will inline it anyways.

File: mono/utils/mono-threads.c [hunk 3]
```diff
@@ -48,7 +48,7 @@ The GC has to acquire this lock before starting a STW to make sure
 a runtime suspend won't make it wronly see a thread in a safepoint
 when it is in fact not.
 */
-static MonoCoopSem global_suspend_semaphore;
+static MonoSemType global_suspend_semaphore;
```
Comment (kumpera) [inline mono/utils/mono-threads.c]: Could you document, for future references, why this semaphore cannot be a MonoCoopSem?

File: mono/metadata/threads.c [hunk 4]
```diff
@@ -4993,9 +4993,12 @@ mono_threads_join_threads (void)
 		}
 		joinable_threads_unlock ();
 		if (found) {
-			if (thread != pthread_self ())
+			if (thread != pthread_self ()) {
```
Comment (kumpera) [inline mono/metadata/threads.c]: Shouldn't we do the same for mono_thread_join ?

File: mono/utils/mono-threads.c [hunk 5]
```diff
@@ -187,30 +187,49 @@ mono_threads_end_global_suspend (void)
...
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x0\t- starting (GOOD, unless the thread is running managed code)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x1\t- running (BAD, unless it's the gc thread)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x2\t- detached (GOOD, unless the thread is running managed code)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x?03\t- async suspended (GOOD)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x?04\t- self suspended (GOOD)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x?05\t- async suspend requested (BAD)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x?06\t- self suspend requested (BAD)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x*07\t- blocking (GOOD)\n");
-	MOSTLY_ASYNC_SAFE_PRINTF ("\t0x?08\t- blocking with pending suspend (GOOD)\n");
+dump_threads (GString *out)
```
Comment (kumpera) [inline mono/utils/mono-threads.c]: What's the reason for this change?
It causes the debug spew to deadlock in the unified suspend case since it mallocs now.

**Comment by kumpera** [comment]
LGTM. All my issues can be addressed after.

File: mono/metadata/threads.c (hunk 4 above)
Comment (ludovic-henry) [inline mono/metadata/threads.c]: done in https://github.com/mono/mono/pull/3074/commits/a9675a64b0d3cd319dc6811be55a220ace3cae49

**Comment by kumpera** [comment]
Anything missing to get this merged?

**Comment by ludovic-henry** [comment]
build
